### PR TITLE
=str #16957 Refactor actor creation api for sinks and sources

### PR DESCRIPTION
**Pull request by patriknw**
- also solves #16952, dispatcher for ActorPublisher

### CODE REVIEW

**Comment by patriknw** [comment]
Refs #16957

File: akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala [hunk 1]
```diff
@@ -19,15 +19,40 @@ import org.reactivestreams._
...
+    val effectiveAttributes: OperationAttributes,
+    val effectiveSettings: ActorFlowMaterializerSettings,
+    val stageName: String,
+    materializerImpl: ActorFlowMaterializerImpl) {
+
+    def materializer: ActorFlowMaterializer = materializerImpl
+
+    def actorOf(props: Props): ActorRef = {
+      val dispatcher =
+        if (props.dispatcher == Dispatchers.DefaultDispatcherId) effectiveSettings.dispatcher
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: you should test for props.dispatcher == NoDispatcherGiven instead. (Look at Props.dispatcher)

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala [hunk 2]
```diff
@@ -4,22 +4,21 @@
...
 import scala.concurrent.{ Future, Promise }
+import akka.stream.impl.ActorFlowMaterializerImpl.CreateContext
 
 /**
  * INTERNAL API
  */
 private[akka] abstract class SinkModule[-In, Mat](val shape: SinkShape[In]) extends Module {
 
-  def create(materializer: ActorFlowMaterializerImpl, flowName: String): (Subscriber[In] @uncheckedVariance, Mat)
+  def create(context: CreateContext): (Subscriber[In] @uncheckedVariance, Mat)
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: this is still tied to ActorFlowMatImpl, but at least it paves the way for generalization

**Comment by drewhk** [comment]
Yes, this is what I imagined. We can make later CreateContext a trait and have an ActorCreateContext for the actorOf method.

File: akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala (hunk 1 above)
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: I thought so too, but actually not:

    def dispatcher: String = deploy.dispatcher match {
      case NoDispatcherGiven ⇒ Dispatchers.DefaultDispatcherId
      case x                 ⇒ x
    }

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: well, you are right (I did this before), check deploy.dispatcher

File: akka-stream/src/main/scala/akka/stream/impl/Sinks.scala (hunk 2 above)
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: yes, I was thinking how far I should go, but stayed here. As far as I can see we don't really support custom sources and sinks anyway, or did I miss anything?

I can make CreateContext independent of actor if you like, but then I think the actorOf method should still be on ActorFlowMaterializer, taking the CreateContext and the Props as parameters.
The source/sink have to downcast to ActorFlowMaterializer when needed.

I think that would be better than CreateContext and ActorCreateContext.

I'll give it a quick try.

Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/Sinks.scala]: no, we should leave it as it is for now. This is enough to allow us later to pull it more apart (that is what I meant to say).

**Comment by patriknw** [comment]
@drewhk I implemented the alternative, anyway, compare and pick. 5fd0cb7

File: akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala [hunk 3]
```diff
@@ -166,13 +176,17 @@ private[akka] case class ActorFlowMaterializerImpl(override val settings: ActorF
...
+  override lazy val executionContext: ExecutionContextExecutor = dispatchers.lookup(settings.dispatcher match {
     case Deploy.NoDispatcherGiven ⇒ Dispatchers.DefaultDispatcherId
     case other                    ⇒ other
   })
 
-  private[akka] def actorOf(props: Props, name: String): ActorRef =
-    actorOf(props, name, settings.dispatcher)
+  override def actorOf(context: MaterializationContext, props: Props): ActorRef = {
+    val dispatcher =
+      if (props.dispatcher == Dispatchers.DefaultDispatcherId) context.effectiveSettings.dispatcher
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: props.deploy.dispatcher == NoDispatcherGiven

If a Props is using the default dispatcher explicitly then we should not override it

**Comment by drewhk** [comment]
I like the second approach better, actually, but I guess others might have a different opinion.

File: akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala (hunk 3 above)
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: ok, good point, will change to that

**Comment by patriknw** [comment]
I also like the second better, if that was not clear :)

File: akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala [hunk 4]
```diff
@@ -121,6 +122,16 @@ object ActorFlowMaterializer {
     system
   }
 
+  /**
+   * INTERNAL API
+   */
+  private[akka] def downcast(materializer: FlowMaterializer): ActorFlowMaterializer =
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala]: good idea :)

File: akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala [hunk 5]
```diff
@@ -121,6 +122,16 @@ object ActorFlowMaterializer {
...
+    materializer match {
+      case m: ActorFlowMaterializer ⇒ m
+      case _ ⇒ throw new IllegalArgumentException(s"required [$classOf[ActorFlowMaterializer.getName]" +
```
Comment (drewhk) [inline akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala]: why the classOf[]?

Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala]: that should be `[${classOf[ActorFlowMaterializer].getName}]`

File: akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala [hunk 6]
```diff
@@ -134,10 +145,12 @@ abstract class ActorFlowMaterializer extends FlowMaterializer {
...
   def settings: ActorFlowMaterializerSettings
 
+  def effectiveSettings(opAttr: OperationAttributes): ActorFlowMaterializerSettings
+
   /**
-   * INTERNAL API
+   * INTERNAL API: this might become public later
    */
-  private[akka] def actorOf(props: Props, name: String): ActorRef
+  private[akka] def actorOf(context: MaterializationContext, props: Props): ActorRef
```
Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala]: No option to provide name? I found `s"$flowName-whichExactlySinkIAm"` useful when looking at "which actor didnt handle a message" when developing things here etc

Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala]: the name should come from the OperationAttributes, which is in the `context`

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/ActorFlowMaterializer.scala]: Okey, makes sense :)

**Comment by ktoso** [comment]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala [hunk 7]
```diff
@@ -43,12 +32,22 @@ private[akka] case class ActorFlowMaterializerImpl(override val settings: ActorF
...
   private[this] def createFlowName(): String = s"$namePrefix-${nextFlowNameCount()}"
 
+  override def effectiveSettings(opAttr: OperationAttributes): ActorFlowMaterializerSettings = {
+    import OperationAttributes._
+    opAttr.attributes.collect {
+      case InputBuffer(initial, max) ⇒ (s: ActorFlowMaterializerSettings) ⇒ s.withInputBuffer(initial, max)
+      case Dispatcher(dispatcher) ⇒ (s: ActorFlowMaterializerSettings) ⇒ s.withDispatcher(dispatcher)
+      case SupervisionStrategy(decider) ⇒ (s: ActorFlowMaterializerSettings) ⇒
+        s.withSupervisionStrategy(decider)
+    }.reduceOption(_ andThen _).getOrElse((x: ActorFlowMaterializerSettings) ⇒ x)(settings) // FIXME is this the optimal way of encoding this?
```
Comment (2m) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: Fold should get rid of this fixme:
``` scala
    opAttr.attributes.foldLeft(settings) { (s, attr) => attr match {
        case InputBuffer(initial, max) ⇒ s.withInputBuffer(initial, max)
        case Dispatcher(dispatcher) ⇒ s.withDispatcher(dispatcher)
        case SupervisionStrategy(decider) ⇒ s.withSupervisionStrategy(decider)
        case Name(_) ⇒ s
      }
    }
```

**Comment by 2m** [comment]
LGTM

File: akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala (hunk 7 above)
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/ActorFlowMaterializerImpl.scala]: ok, I'll fix that

**Comment by patriknw** [comment]
squashed, will merge after pr validation

**Pull request merged**